Improve position named arguments lint underline and formatting names

PrestonFrom opened the pull request:
For named arguments used as implicit position arguments, underline both
the opening curly brace and either:
* if there is formatting, the next character (which will either be the
  closing curl brace or the `:` denoting the start of formatting args)
* if there is no formatting, the entire arg span (important if there is
  whitespace like `{  }`)

This should make it more obvious where the named argument should be.

Additionally, in the lint message, emit the formatting argument names
without a dollar sign to avoid potentially confusion.

Fixes #99907

rust-highfive commented:
r? @michaelwoerister

(rust-highfive has picked a reviewer for you, use r? to override)

PrestonFrom commented:
@lqd This addresses your comment here: https://github.com/rust-lang/rust/pull/99660#discussion_r932952838

Not sure if you want to be the reviewer here or not though?

compiler-errors commented:
r? @compiler-errors since i reviewed the other two PRs touching this code

compiler-errors commented on src/test/ui/macros/issue-99265.stderr:
```diff
@@ -190,7 +190,7 @@ LL |         "{}, Hello {1:2$.3$} {4:5$.6$}! {1}",
    |                     - this formatting argument uses named argument `f` by position
 ...
 LL |         f = 0.02f32,
-   |         ^ this named argument is only referred to by position in formatting string
```
Did you mean to drop "only" here?

PrestonFrom replied:
With the previous PR, the lint will trigger for every case a named argument is used positionally even if it's used by name elsewhere, so I think "only" is now incorrect. (I missed this in the last change.)

compiler-errors replied:
Ah, ok.

compiler-errors commented on src/test/ui/macros/issue-99265.stderr:
```diff
@@ -16,9 +16,9 @@ warning: named argument `a` is not used by name
   --> $DIR/issue-99265.rs:9:35
    |
 LL |     println!("{} {} {} {} {}", 0, a=1, b=2, c=3, d=4);
-   |                   -               ^ this named argument is only referred to by position in formatting string
```
Is the loss of "only" intentional?

compiler-errors replied:
duplicate lol, sorry that last note didnt send

compiler-errors commented on compiler/rustc_builtin_macros/src/format.rs:
```diff
@@ -57,26 +58,47 @@ struct PositionalNamedArg {
-                .arg_spans
-                .get(self.cur_piece)
-                .map(|arg_span| arg_span.with_lo(arg_span.lo() + BytePos(1)).shrink_to_lo());
+            // In the case of a named argument whose position is implicit, if the argument *has*
```
Thanks for the detailed comment.

compiler-errors commented on compiler/rustc_builtin_macros/src/format.rs:
```diff
@@ -57,26 +58,47 @@ struct PositionalNamedArg {
+            // For the message span, if there is formatting, we want to use the opening `{` and the
+            // next character, which will the `:` indicating the start of formatting. If there is
+            // not any formatting, we want to underline the entire span.
+            if self.has_formatting {
```
Is it possible to factor this into the body of the closure? Both the `if` and `else` have some duplication, so pushing the `if` into the closure might deduplicate some.

PrestonFrom replied:
That's a great idea!

PrestonFrom commented:
@compiler-errors Thank you for the review!

It looks like #99987 is still in the queue but at the moment I don't think it should have any merge conflicts.

compiler-errors commented:
Alright, seems like it didn't conflict.

@bors r+

bors commented:
:pushpin: Commit 298acef30730ee676fdbc9e731370cd7ccedd431 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 298acef30730ee676fdbc9e731370cd7ccedd431 -->
<!-- homu: {"type":"Approved","sha":"298acef30730ee676fdbc9e731370cd7ccedd431","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
